File: autologger/autologger.go
package autologger

// This package is one implementation to control what to be logged for each API automatically.
// It provides customization function to go-grpc-middleware logging interceptor.

import (
	"context"
	"fmt"

	log "log/slog"

	"github.com/Azure/aks-middleware/requestid"

	"github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors/logging"
)

func InterceptorLogger(logger *log.Logger) logging.Logger {
	return logging.LoggerFunc(func(ctx context.Context, lvl logging.Level, msg string, fields ...any) {
		// fmt.Println("ctx: ", ctx)
		// fmt.Printf("fields: %v\n", fields)
		f := make(map[string]any, len(fields)/2)
		l := logger
		i := logging.Fields(fields).Iterator()
		for i.Next() {
			k, v := i.At()
			f[k] = v
			l = l.With(k, v)
			// fmt.Printf("k %v, v %v\n", k, v)
		}

		// fmt.Println(lvl, msg)
		// l.Info("blah")

		switch lvl {
		case logging.LevelDebug:
			l.Debug(msg)
		case logging.LevelInfo:
			l.Info(msg)
		case logging.LevelWarn:
			l.Warn(msg)
		case logging.LevelError:
			l.Error(msg)
		default:
			panic(fmt.Sprintf("unknown level %v", lvl))
		}
	})
}

// Add request-id & headers to API autologger.
func GetFields(ctx context.Context) logging.Fields {
	headers := requestid.GetRequestHeaders(ctx)
	requestID := headers[requestid.RequestIDMetadataKey]
	// Remove the main request ID from headers map since it's logged separately
	delete(headers, requestid.RequestIDMetadataKey)
	return logging.Fields{
		requestid.RequestIDLogKey, requestID,
		"headers", headers,
	}
}

File: autologger/autologger.cs
using System;
using Grpc.Core;
using Grpc.Core.Interceptors;
using Microsoft.Extensions.Logging;

namespace Azure.Aks.Middleware.Autologger
{
    public class AutoLoggerInterceptor : Interceptor
    {
        private readonly ILogger _logger;

        public AutoLoggerInterceptor(ILogger<AutoLoggerInterceptor> logger)
        {
            _logger = logger;
        }

        public override async Task<TResponse> UnaryServerHandler<TRequest, TResponse>(
            TRequest request,
            ServerCallContext context,
            UnaryServerMethod<TRequest, TResponse> continuation)
        {
            try
            {
                LogCall(context.Method, LogLevel.Information, "Starting call.");
                var response = await continuation(request, context);
                LogCall(context.Method, LogLevel.Information, "Call finished successfully.");
                return response;
            }
            catch (Exception ex)
            {
                LogCall(context.Method, LogLevel.Error, $"Call threw an exception: {ex.Message}");
                throw;
            }
        }

        private void LogCall(string methodName, LogLevel logLevel, string message)
        {
            _logger.Log(logLevel, $"{methodName}: {message}");
        }
    }
}
